osquery/tables/system/darwin/asl_utils.cpp

```diff
@@ -0,0 +1,108 @@
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+
+#include "asl_utils.h"
+
+namespace ba = boost::algorithm;
+
+namespace osquery {
+namespace tables {
+
+bool isNumeric(ColumnType coltype) {
+  switch (coltype) {
+  case INTEGER_TYPE:
+  case BIGINT_TYPE:
+  case UNSIGNED_BIGINT_TYPE:
+  case DOUBLE_TYPE:
+    return true;
+  default:
+    return false;
+  }
+}
+
+void addQueryOp(asl_object_t& query,
+                const std::string& key,
+                const std::string& value,
+                ConstraintOperator op,
+                ColumnType col_type) {
+  if (key == kExtraColumnKey) {
+    // ASL doesn't know about the 'Extra' column, so we can't do the matching
+    // through the ASL query. Do nothing here, and later SQLite's engine will
+    // do the match.
+    return;
+  }
+
+  // Only some queries can be supported through ASL, those that are not will
+  // just be performed later by the SQLite engine.
+  auto it = kSupportedAslOps.find(op);
+  if (it != kSupportedAslOps.end()) {
+    std::string modified_val = value;
+    uint32_t asl_op = it->second;
+    switch (op) {
+    case LIKE:
+      modified_val = convertLikeRegex(value);
+      break;
+    default:
+      if (isNumeric(col_type)) {
+        asl_op |= ASL_QUERY_OP_NUMERIC;
+      }
+    }
+    asl_set_query(query, key.c_str(), modified_val.c_str(), asl_op);
+  }
+}
+
+asl_object_t createAslQuery(const QueryContext& context) {
+  asl_object_t query = asl_new(ASL_TYPE_QUERY);
+
+  // Set constraints in query
+  for (const auto& it : context.constraints) {
+    const std::string& key = it.first;
+    ColumnType col_type = it.second.affinity;
+    for (const auto& constraint : it.second.getAll()) {
+      addQueryOp(query, key, constraint.expr, static_cast<ConstraintOperator>(constraint.op), col_type);
+    }
+  }
+  return query;
+}
+
+void readAslRow(asl_object_t row, Row& r, const QueryContext& context) {
+  pt::ptree extras;
+  int num_columns = asl_count(row);
+
+  // Fetch each column individually, adding it to the result map
+  for (int i = 0; i < num_columns; ++i) {
+    const char *key, *val;
```

Comment:
Let's initialize these to `nullptr` <--- it helps A LOT with static analysis.
 `make analyze` will run LLVM's static analysis tools and error on warnings.